### PR TITLE
feat!: integrate cosmwasm

### PR DESCRIPTION
**Pull request by reecepbcups**
ref: #3050

Adds CosmWasm and makes it permissioned in the upgrade handler per [proposal 895](https://www.mintscan.io/cosmos/proposals/895) signaling to add. 

# Test Local node Script
```
# The testnode here is permissionless. On upgrade handler, permissioned is set
GAIAD_NODE=http://localhost:26657
rm -rf ~/.gaia/ && make install
sh contrib/single-node.sh local-1 cosmos10r39fueph9fq7a6lgswu4zdsg8t3gxlqvvvyvn
```

## Upload Wasm Contract
```
sh ./contrib/upload-contract.sh
```

## Local Upgrade
```
go install github.com/cosmos/cosmos-sdk/cosmovisor/cmd/cosmovisor@latest
GAIAD_NODE=http://localhost:26657

git checkout v15.1.0
make build
cp ./build/gaiad ./build/gaiadold
go clean -modcache

git checkout wasmd-integration
make build
cp ./build/gaiad ./build/gaiadnew

./contrib/scripts/upgrade_test_scripts/run_gaia.sh

./contrib/scripts/upgrade_test_scripts/run_upgrade_commands.sh 15

./contrib/scripts/upgrade_test_scripts/test_upgrade.sh 20 5 16 localhost

gaiad q wasm params
```

# TODO:
- [x] Fix Upgrade CI (which works locally)
- [ ] Changelog

---

image

### CODE REVIEW

File: app/upgrades/v16/upgrades.go
Comment (reecepbcups) [inline app/upgrades/v16/upgrades.go]: Need input on the team for if this. Should it be 
- `InstantiateDefaultPermission = AccessTypeAnyOfAddresses`  _(whitelisted wallets can instantiate new contracts as much as they want off of approve code ids from gov)_
- or `InstantiateDefaultPermission = AccessTypeNobody` _(only gov only can create contracts off previously gov approved code)_

To me `AccessTypeAnyOfAddresses` makes sense.

File: app/upgrades/v16/upgrades.go
Comment (reecepbcups) [reply app/upgrades/v16/upgrades.go]: false, used in keepers.WasmKeeper.SetParams(ctx, wasmParams) & confirmed in a local upgrade

image

File: app/upgrades/v16/upgrades.go
Comment (reecepbcups) [reply app/upgrades/v16/upgrades.go]: same as below

**Comment by reecepbcups** [comment]
Upgrade works locally with versions but not here, odd. 
image

File: .goreleaser.yml
Comment (webmaster128) [inline .goreleaser.yml:26]: There should not be anything wrong with doing the same for ARM. I'd just not recommend running on ARM CPUs for production systems as Singlepass still has some smaller issues on ARM.

File: .goreleaser.yml
Comment (webmaster128) [inline .goreleaser.yml:23]: On Windows we can still build a client binary (through some build flags). However, the contract execution on Windows is not supported.

File: Dockerfile
Comment (webmaster128) [inline Dockerfile]: I'd call this one `WASMVM_VERSION` as this is really versioning the wasmvm repo

File: Dockerfile
Comment (webmaster128) [inline Dockerfile]: Shoudn't this fail hard in the else case?

File: Dockerfile
Comment (webmaster128) [inline Dockerfile]: This blindly trusts whatever GitHub returns. I think at least the checksum check should be added here even it adds a tiny bit of upgrade work. The other option would be signatures but we don't have a setup for this right now.

File: Dockerfile
Comment (webmaster128) [inline Dockerfile]: [We currently need a build system with Alpine < 3.19](https://github.com/CosmWasm/wasmvm/issues/523). This seems to do the job for most teams:

```suggestion
FROM golang:1.21-alpine3.18 AS gaiad-builder
```

File: app/keepers/keepers.go
Comment (webmaster128) [inline app/keepers/keepers.go]: Any reason for not using `AllCapabilities()` which returns exactly that list? I mean yeah, it has a misleading doc comment and is in the wrong package. But we'll fix those minor flaws. The function itself is good IMO as it simply returns all capabilities that the current wasmd has.

File: e2e.Dockerfile
Comment (webmaster128) [inline e2e.Dockerfile]: Same points as in the other Dockerfile

**Review by webmaster128 — COMMENTED** [review]
Great stuff 🎉 I'm sure you know the Go integration part better than myself

File: app/keepers/keepers.go
Comment (reecepbcups) [reply app/keepers/keepers.go]: I really dislike having to use the wasmd app namespace to import this type. This was why I originally wanted it in wasmvm. tbh I think it makes sense to put in the wasmd/x/wasm/types as a const or something. Maybe I am being too picky idk

Will use `wasmapp.AllCapabilities()`

File: Dockerfile
Comment (reecepbcups) [reply Dockerfile]: Good call out, I actually forgot to  do the static link build tag in the make file. So in the event this fails, it will fail building and give the correct output that it could not find libwasmvm. Is nice because it keeps bash to a minimum

File: .goreleaser.yml
Comment (reecepbcups) [reply .goreleaser.yml:26]: Re-enabled in e9058051f89db82238d6b2e17eccaa41e39adcb5

File: Dockerfile
Comment (reecepbcups) [reply Dockerfile]: e9058051f89db82238d6b2e17eccaa41e39adcb5

File: Dockerfile
Comment (reecepbcups) [reply Dockerfile]: ref makefile e9058051f89db82238d6b2e17eccaa41e39adcb5

File: app/keepers/keepers.go
Comment (reecepbcups) [reply app/keepers/keepers.go]: e9058051f89db82238d6b2e17eccaa41e39adcb5

File: .goreleaser.yml
Comment (reecepbcups) [reply .goreleaser.yml:23]: I'm not sure if goreleaser lets you specify the `GOOS=windows GOARCH=amd64 go build $(BUILD_FLAGS) -o build/gaiad.exe ./cmd/gaiad` for just 1 type of arch. Will have to look more into

File: Dockerfile
Comment (reecepbcups) [reply Dockerfile]: resolved: 0f5ebcdf0afd3b6e8ecc83b24a99556326a87c93

File: contrib/single-node.sh
Comment (reecepbcups) [inline contrib/single-node.sh:29]: These scripts probably need to be redone tbh. Happy to do if the team would like. Maybe another PR

ref what I use: https://github.com/CosmosContracts/juno/blob/main/scripts/test_node.sh

**Comment by MSalopek** [comment]
@Reecepbcups 
Hey, thanks for the your contribution! Clearly, there was quite a lot of work that went into this and we appreciate the effort!

I'd like to help pass all the checks (upgrade tests, e2e etc.) if you don't mind me adding to the PR?

As for merging this, the v16 release is already feature frozen and will be cut tomorrow morning (European time) as we have already prepared all the testing pipelines and made plans for the testnet upgrades. It's a pretty beefy release adding IBC features voted in by governance.

I think it should be pretty safe to merge this into main once we finalize the v16 release branch.

I'm really excited about this :rocket:

File: .goreleaser.yml
Comment (MSalopek) [reply .goreleaser.yml:23]: I can take a look at this but I think you're right about the env vars.

File: contrib/single-node.sh
Comment (MSalopek) [reply contrib/single-node.sh:29]: We have this in the backlob but we haven't started refactoring yet: https://github.com/cosmos/gaia/issues/2950

We can add refactoring this script to that issue and do them all in a different PR.

**Comment by reecepbcups** [comment]
@MSalopek Thanks and Sounds good on the v16 front. Yes, Maintainer edits are enabled.

**Comment by reecepbcups** [comment]
I’ll pull in main Tuesday in prep for v18*

**Comment by MSalopek** [comment]
> I’ll pull in main Tuesday in prep for v17

I already merged main into the branch (maintainer edits are enabled on the `wasm-integration`  branch). I'm fixing tests.

v17 is already cut and the propsal is in voting (it has partial set security features enabled with some other goodies).
This brings a major change for the ICS protocol and all consumer chains so it was made into its own release.

v18 is next with cosmwasm, ibc-go v7.5.0 and some other upgrades. The PR will be merged as soon as all tests are passing and we provide a definite answer regarding this discussion: https://github.com/cosmos/gaia/pull/3051#discussion_r1558435243.

The relatively fast upgrade cadence enables us to push out more features in distinct releases so we offset some risks around the upgrade.

File: app/upgrades/v16/upgrades.go
Comment (MSalopek) [reply app/upgrades/v16/upgrades.go]: We want the governance to handle both code upload and instantiation. 

IIUC, to use permissiond wasm we would need to use these settings:
```golang
wasmParams := wasmtypes.Params{
	CodeUploadAccess:             wasmtypes.AllowNobody,
	InstantiateDefaultPermission: wasmtypes.AllowNobody,
}
```

~~I don't think that code upload is followed by an instantiation operation by default. But, that can be can be achieved through a gov prop that has 2 Txs to execute: `[MsgStoreCode, MsgInstantiateContract]`. This way, the uploaded contracts would immediately get instantiated.~~

We can use `MsgStoreAndInstantiateContract` to store and instantiate a contract via gov proposal.

Do you have any comments on this @Reecepbcups ?

File: e2e.Dockerfile
Comment (MSalopek) [inline e2e.Dockerfile:20]: Had issues with this in e2e.

CGO needs to be enabled (don't know why).
I've also had to add `RUN apk add --no-cache build-base
` in the runner container to make CGO available.

It seems that we cannot run the tests without CGO enabled.

I'm not sure what the implications of this are. While checking other projects, they all seem to be running with CGO enabled (or at least they don't explicitly disable it using `CGO_ENABLED=0`.

Comment (reecepbcups) [reply e2e.Dockerfile:20]: > This repo contains both Rust and Go code. The Rust code is compiled into a library (shared .dll/.dylib/.so or static .a) to be linked via cgo and wrapped with a pleasant Go API

CGO is a requirement for CW yes https://github.com/CosmWasm/wasmvm

File: app/upgrades/v16/upgrades.go
Comment (reecepbcups) [reply app/upgrades/v16/upgrades.go]: This makes total sense yes, wasmtypes.AllowNobody is the desired param for both

[MsgStoreAndInstantiateContract](https://github.com/CosmWasm/wasmd/blob/v0.45.0/proto/cosmwasm/wasm/v1/proposal_legacy.proto#L310) looks like it will allow the uploader (gov) to override the AccessConfig though for instantiates.  `InstantiateContractProposal` does not have this. 

I would assume this overrides the defaults since gov approved, which is a feature but actually a bug in our case for expected behavior. i.e. need to ante block a proposal if `MsgStoreAndInstantiateContract.instantiate_permission` is set

Comment (MSalopek) [reply app/upgrades/v16/upgrades.go]: I think I'm getting it, thanks!

We will address those concerns before testnet and write new docs before cosmwasm lands on the hub mainnet.

**Review by MSalopek — APPROVED** [review]
Thanks again for the PR!

All tests were cleaned up, the upgrade is passing.

There's a bit of work to do later.
We still need to check the goreleaser settings, update Dockerfile and make better upgrade tests.

The build got a bit complicated given the `libwasmvm` dep that can be either statically or dynamically linked. `upgrade-test` is using a dynamic linking approach while dockerized build uses static linking.

We will need to expand the build docs to show how to build with cosmwasm in mind.
* https://github.com/cosmos/gaia/issues/3119

File: app/upgrades/v16/upgrades.go
Comment (dasanchez) [reply app/upgrades/v16/upgrades.go]: Hi! Do you have an example of what a proposal that uses `MsgStoreAndInstantiateContract` would look like? I would like to know how the `wasm_byte_code` field must be encoded:
```
{
 "messages": [
  {
   "@type": "/cosmwasm.wasm.v1.MsgStoreAndInstantiateContract",
   "authority": "cosmos10d07y265gmmuvt4z0w9aw880jnsr700j6zn9kn",
   "wasm_byte_code": null,  <- My understanding is the .wasm file contents will go in here
   "instantiate_permission": null,
   "unpin_code": false,
   "admin": null,
   "label": "label",
   "msg": null,
   "funds": [],
   "source": null,
   "builder": null,
   "code_hash": null
  }
 ],
 "metadata": "ipfs://CID",
 "deposit": "10000000uatom",
 "title": "Test Proposal for Store and Instantiate",
 "summary": "none"
}
```

Comment (reecepbcups) [reply app/upgrades/v16/upgrades.go]: @dasanchez It can be raw or gzip compressed from the source file. Base64 encoded

So like `cat artifacts/clock_example.wasm | base64 -w 0` or `cat artifacts/clock_example.wasm | gzip -cf | base64 -w 0` iirc

Comment (dasanchez) [reply app/upgrades/v16/upgrades.go]: @Reecepbcups thanks! I ended up using the `submit-proposal` command from the wasm module (rather than the gov draft-proposal template) and managed to 1) store and instantiate, and 2) execute via proposals:
```
gaiad tx wasm submit-proposal store-instantiate my_first_contract.wasm '{"count":100}' \
    --label "my first contract" \
    --no-admin \
    --instantiate-nobody true \
    --title "Store and instantiate CW template" \
    --summary "This proposal will store and instantiate the cw template contract"
```

```
gaiad tx wasm submit-proposal execute-contract $contract_address \
  '{"increment":{}}' --title "Execute CW template" \
   --summary "This proposal will execute an increment on the cw template contract"
```

```
gaiad q wasm contract-state smart $contract_address '{"get_count":{}}'
data:
  count: 101
```

File: ante/ante.go
Comment (dusan-maksimovic) [inline ante/ante.go]: Looks like `WasmKeeper` isn't used anywhere, can we remove it?

File: app/keepers/keepers.go
Comment (dusan-maksimovic) [inline app/keepers/keepers.go:484]: Can't add a comment on line 513 since it wasn't changed. We need to wire-up the wasm module with IBC stack, otherwise smart contracts will not be able to create channels etc. See the snippet (and figure out if we want fee in the stack):

```go
var wasmStack porttypes.IBCModule
	wasmStack = wasm.NewIBCHandler(app.WasmKeeper, app.IBCKeeper.ChannelKeeper, app.IBCFeeKeeper)
	wasmStack = ibcfee.NewIBCMiddleware(wasmStack, app.IBCFeeKeeper)
...
ibcRouter := porttypes.NewRouter().
     AddRoute(wasmtypes.ModuleName, wasmStack).
```

Comment (reecepbcups) [reply app/keepers/keepers.go:484]: resolved: 56c28cd5efc480a4af1375ed9770010ad8411b59

File: ante/ante.go
Comment (reecepbcups) [reply ante/ante.go]: resolved: b22dcd3859df2495b4e168c1a5566bce80c461f9

**Review by stana-miric — APPROVED** [review]

File: app/modules.go
Comment (stana-miric) [inline app/modules.go:1]: Do we need to update maccPerms with wasm module?

Comment (MSalopek) [reply app/modules.go:1]: I added this permission based on `wasmd` [example app](https://github.com/CosmWasm/wasmd/blob/main/app/app.go).
```golang
wasmtypes.ModuleName:              {authtypes.Burner},
```

We can remove it if it interferes with something.

**Comment by reecepbcups** [comment]
:fire:

**Pull request merged**
